Automattic/newspack-blocks
# Add autocomplete search to article block author/tag/category inputs

### All Submissions:






### Changes proposed in this Pull Request:



The REST API limit for fetching a bunch of stuff is 100 items. Virtually all large sites have >100 users and >100 tags, so the current select-dropdown-based author/category/tag select will not be usable on a real site. 

To address this problem, I've created and added an autocompleting search component for those fields. 

The fields in various states look like this:

<img width="274" alt="Screen Shot 2019-10-01 at 12 04 44 PM" src="https://user-images.githubusercontent.com/7317227/65992308-ca54af00-e443-11e9-8e41-bb3a3f329cd0.png">

### Component

I've introduced a new `AutocompleteTokenField` component to handle all of the heavy lifting. This component should be fairly general-purpose, so it can be used in @georgeh's query block work if needed. One important thing this does compared to the regular `FormTokenField` is enforce known values; you can't e.g. just create new tags like with the normal one.

I designed it primarily to be super-usable for block attributes. You just pass in the raw attributes and two functions that convert those attributes into objects with `label` and `value` properties:

```
// Convert a string into autocomplete results.
const fetchTagSuggestions = ( search ) => {
	return apiFetch( {
		path: addQueryArgs( '/wp/v2/tags', {
			search,
			per_page: 20,
			_fields: 'id,name',
			orderby: 'count',
			order: 'desc',
		} ),
	} ).then( function( tags ) {
		return map( tags, ( tag ) => ( {
			value: tag.id,
			label: decodeEntities( tag.name ) || __( '(no title)' ),
		} ) );
	} );
};
// Convert saved attribute value into recognizable data.
const fetchSavedTags = ( tagIDs ) => {
	return apiFetch( {
		path: addQueryArgs( '/wp/v2/tags', {
			per_page: 100,
			_fields: 'id,name',
			include: tagIDs.join(','),
		} ),
	} ).then( function( tags ) {
		return map( tags, ( tag ) => ( {
			value: tag.id,
			label: decodeEntities( tag.name ) || __( '(no title)' ),
		} ) );
	} );
};

<AutocompleteTokenField
	tokens={ tags || [] }
	onChange={ ( tokens ) => setAttributes( { tags: tokens } ) }
	fetchSuggestions={ fetchTagSuggestions }
	fetchSavedInfo={ fetchSavedTags }
	label={ __( 'Tag' ) }
/>
```

### How to test the changes in this Pull Request:

1. `npm run build:webpack`
2. Insert an Article block. Try out the fields and verify they work as described above.
3. Save the page. Verify the fields save and load correctly on a saved page. Verify the fields filter correctly on the site frontend.

### Other information:
